typing/persistent_env.ml

```diff
@@ -0,0 +1,344 @@
+(* Persistent structure descriptions *)
+
+open Misc
+open Cmi_format
+
+let add_delayed_check_forward = ref (fun _ -> assert false)
+
+type error =
+  | Illegal_renaming of modname * modname * filepath
+  | Inconsistent_import of modname * filepath * filepath
+  | Need_recursive_types of modname
+  | Depend_on_unsafe_string_unit of modname
+
+exception Error of error
+let error err = raise (Error err)
+
+module Persistent_signature = struct
+  type t =
+    { filename : string;
+      cmi : Cmi_format.cmi_infos }
+
+  let load = ref (fun ~unit_name ->
+      match Load_path.find_uncap (unit_name ^ ".cmi") with
+      | filename -> Some { filename; cmi = read_cmi filename }
+      | exception Not_found -> None)
+end
+
+type can_load_cmis =
+  | Can_load_cmis
+  | Cannot_load_cmis of EnvLazy.log
+
+type pers_struct = {
+  ps_name: string;
+  ps_crcs: (string * Digest.t option) list;
+  ps_filename: string;
+  ps_flags: pers_flags list;
+}
+
+module String = Misc.Stdlib.String
+
+type 'a t = {
+  persistent_structures : (string, (pers_struct * 'a) option) Hashtbl.t;
```

Comment:
Maybe we should use a specific datatype here. I didn't immediately remember what `None` meant. For instance:

```ocaml
type 'a pers_struct_info =
  | Not_found
  | Found of { pers_struct : pers_struct; data : 'a }

type 'a t = {
  persistent_structures : (string, 'a pers_struct_info) Hashtbl.t;
...
}
```